docs/api.rst

```diff
@@ -3,8 +3,72 @@ Python API
 
 Pipeline
 --------
+In DALI, any data processing task has a central object called pipeline. Pipeline object is an
+instance of :class:`nvidia.dali.pipeline.Pipeline` or a derived class. Pipeline encapsulates the
+data processing graph and the execution engine.
+
+There are two ways to define a DALI pipelines:
+
+#. by inheriting from Pipeline class and overriding :meth:`nvidia.dali.pipeline.Pipeline.define_graph`
+#. by instantiating `Pipeline` directly, building the graph and setting the pipeline
+   outputs with :meth:`nvidia.dali.pipeline.Pipeline.set_outputs`
+
+Data processing graphs
+""""""""""""""""""""""
+
+DALI pipeline is represented as a graph of operations. There are two kinds of nodes in the graph:
+
+ * Operators
+ * Data nodes (see :class:`nvidia.dali.pipeline.DataNode`)
+
+The nodes are created on each call to an operator object (or function) - each call typically
+creates one operator node and as many data nodes as there are outputs from the operator.
+Passing `DataNode` objects as inputs or arguments to an operator establishes connections in
+the graph.
```

Comment:
This sounds complicated.
Can we maybe break that down a little, and not mix both approaches?

Not sure if it's any better.
```suggestion
DALI pipeline is represented as a graph of operations. There are two kinds of nodes in the graph:

 * Operators - represent the computations and are created by invoking the call `()` on operator object or using the fn.operator wrapper.
 * Data nodes (see :class:`nvidia.dali.pipeline.DataNode`) - they represent the data exchanged between operators. The `DataNode` are returned from the call to Operator object indicating its outputs and can be passed to another operator as its input to establish connections in the graph.
```

Reply:
I'll rephrase it - but certainly not with "invoking the call ()", which doesn't seem right either - `()` is a C++ thing - i specifically didn't refer to __call__ for future compatibility with functional API and () suggests no arguments.

Reply:
Just an example with how to approach this. I have no good way of writing about the `__call__`, as it may not be obvious for some people what it actually is, and writing about `operator call of the DALI Operator` also brings mostly confusion. :v